### PR TITLE
fix: Fix bad assumptions in Home screen callout selection and refactor for clarity.

### LINKED ISSUE #676: App crashes when I attempt to search for sites with non-matching keyword
### App Version

59

### Account

_No response_

### Platforms

- [X] iOS
- [X] Android
- [ ] iOS simulator
- [ ] Android emulator

### Description

App crashes when I type in a search term that does not match a site or site's notes

### Steps to Reproduce

1. Create a site
2. Open the site and take a note
3. Go back to the sites list and search for the search term
4. Change the search term to something you know doesn't match the site

### Expected behavior

When you type in a search term that is in the site's notes, the site should be returned. When you type in a search term that isn't in the site's notes, the app should tell you there are no matching sites.

### Actual behavior

The matching search term case works fine. The non-matching search term causes the app to crash on both my Android and iPad.

### Additional context

https://github.com/techmatters/terraso-mobile-client/assets/2730009/1f4c1779-85ce-4fd6-88ac-28f3f22b74c9

### PR DESCRIPTION
**Pull request by tm-ruxandra**
## Description

If the user's current callout selection is excluded by their site filters, the app gracefully hides the callout rather than crashing. To enable this fix, add helper functions to the home screen's callout system to make initializing state more concise and mapping state to logic more consistent, and further refactor SiteMapCallout for general clarity.

This PR will fix the three below tickets, but several other bugs are also fixed (similar issues would have affected site clusters and filtering.) The general problem was that components which rendered based on the callout state were not written to consider that the selected callout site(s) might be excluded by the filter.

Also do some general incremental cleanup to the Home screen for readability and clarity: relocate filters and callout state to their own files, and remove the unused `showCallout`state property.

### Checklist
- [x] Corresponding issue has been opened

### Related Issues
Fixes #676, #953, #741 

### Verification steps

Select a site in the home screen. Enter any search or filter criteria which exclude the site, and observe that the application no longer crashes. Do the same for site clusters. Verify that location selections are unaffected.

### CODE REVIEW

File: dev-client/src/screens/HomeScreen/components/SiteMap.tsx
Comment (tm-ruxandra) [inline dev-client/src/screens/HomeScreen/components/SiteMap.tsx:193]: comment: Because the `showCallout` property was unused, this callback could be safely eliminated entirely.

File: dev-client/src/screens/HomeScreen/components/SiteMapCallout.tsx
Comment (tm-ruxandra) [inline dev-client/src/screens/HomeScreen/components/SiteMapCallout.tsx:70]: comment: I made an arbitrary call here on functionality: if the user's selected site is not available in filtered data, the callout will be invisible until the filter includes it again. (The same happens for an empty list of cluster sites). We could also simply invoke the `closeCallout` function to clear the user's selection, but this approach felt more user-friendly.

**Review by shrouxm — APPROVED** [review]
LGTM, no blocking comments, and i gave it a test drive and can't get it to crash!

**Review by paulschreiber — COMMENTED** [review]
LGTM. Will wait for approval from Garo or Carissa.

File: dev-client/src/screens/HomeScreen/HomeScreenCallout.ts
Comment (tm-ruxandra) [reply dev-client/src/screens/HomeScreen/HomeScreenCallout.ts:45]: Just a habit of mine, no particular reason

File: dev-client/src/screens/HomeScreen/HomeScreenCallout.ts
Comment (tm-ruxandra) [reply dev-client/src/screens/HomeScreen/HomeScreenCallout.ts:35]: Not at all, I was already unsure if I was adding too much complexity and your comment gave me the nudge I needed to prefer a more simple approach that was more similar to what already exists in the codebase.

File: dev-client/src/screens/HomeScreen/HomeScreenCallout.ts
Comment (tm-ruxandra) [reply dev-client/src/screens/HomeScreen/HomeScreenCallout.ts:4]: Fixing, good catch

**Pull request merged**
